src/library/scala/concurrent/impl/Promise.scala

```diff
@@ -294,14 +294,21 @@ private[concurrent] final object Promise {
     }
 
     // IMPORTANT: Noop should not be passed in here
-    private[this] final def submitWithValue(callbacks: Callbacks[T], resolved: Try[T]): Unit = {
-      if (callbacks.isInstanceOf[ManyCallbacks[T]]) {
-        val m = callbacks.asInstanceOf[ManyCallbacks[T]]
-        if (m.first.isInstanceOf[Transformation[T, _]]) m.first.asInstanceOf[Transformation[T, _]].submitWithValue(resolved)
-        else submitWithValue(m.first, resolved) // FIXME this will grow the stack
-        submitWithValue(m.last, resolved)
-      } else /*if (callbacks ne Noop)*/ {
-        callbacks.asInstanceOf[Transformation[T, _]].submitWithValue(resolved)
+    @tailrec
+    private[this] final def submitWithValue(callbacks: Callbacks[T],
+                                            resolved: Try[T],
+                                            deferred: List[Callbacks[T]] = Nil): Unit = {
+      callbacks match {
+        case m: ManyCallbacks[T] =>
+          val first = m.first
+          val last = m.last
+          submitWithValue(first, resolved, last :: deferred)
```

Comment:
Do I need change it to `submitWithValue(m.first, resolved, m.last :: deferred)`?

Reply:
`submitWithValue(m.first, resolved, m.last :: deferred)` seems more concise to me.

Reply:
@satansk I have addressed it, thanks.

Reply:
I agree, making it a oneliner makes a lot of sense.